ec2/utils.py

```diff
@@ -76,120 +215,267 @@ def get_instance_state(instance, ctx):
+                                  .format(str(e)))
+    if len(group) is not 1:
+        report_all_security_groups(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id or name: '
+                                  'Too many groups returned.')
 
+    return group
 
-def save_key_pair(key_pair_object, ctx):
-    """ Saves the key pair to the file specified in the blueprint
-    """
 
-    ctx.logger.debug('Attempting to save the key_pair_object.')
+def get_key_pair_by_id(ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
 
     try:
-        key_pair_object.save(ctx.node.properties['private_key_path'])
-    except OSError:
-        raise NonRecoverableError('Unable to save key pair to file: {0}.'
-                                  'OS Returned: {1}'.format(
-                                      ctx.node.properties['private_key_path'],
-                                      OSError))
+        key = ec2_client.get_key_pair(ctx.node.properties['resource_id'])
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to get key pair: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+
+    return key.name
+
+
+def get_address_by_id(address_id, ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+    try:
+        address = ec2_client.get_all_addresses(address_id)
+    except boto.exception.EC2ResponseError as e:
+        report_all_addresses(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except (boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    return address.public_ip
+
+
+def get_address_object_by_id(address_id, ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+    try:
+        address = ec2_client.get_all_addresses(address_id)
+    except boto.exception.EC2ResponseError as e:
+        report_all_addresses(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except (boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    return address
+
+
+def report_all_instances(ctx):
```

Comment:
Changed to get_all_instances and if it fails it runs 'log_available_resources' and then raises non-recoverable